tools/eslint-plugin/dist/rules/public-extension-exports.js

```diff
@@ -0,0 +1,240 @@
+*--------------------------------------------------------------------------------------------*/
+
+// Writing all of these eslint rules in javascript so we can run them before the build step
+
+"use strict";
+
+const { getParserServices } = require("./utils/parser");
+const ts = require("typescript");
+const fs = require("fs");
+
+const syntaxKindFriendlyNames = {
+  [ts.SyntaxKind.ClassDeclaration]: "class",
+  [ts.SyntaxKind.EnumDeclaration]: "enum",
+  [ts.SyntaxKind.InterfaceDeclaration]: "interface",
+  [ts.SyntaxKind.ModuleDeclaration]: "module",
+  [ts.SyntaxKind.MethodDeclaration]: "method",
+  [ts.SyntaxKind.MethodSignature]: "method",
+  [ts.SyntaxKind.FunctionDeclaration]: "function",
+  [ts.SyntaxKind.PropertyDeclaration]: "property",
+  [ts.SyntaxKind.PropertySignature]: "property",
+  [ts.SyntaxKind.Constructor]: "constructor",
+  [ts.SyntaxKind.EnumMember]: "enum member",
+  [ts.SyntaxKind.TypeAliasDeclaration]: "type alias",
+  [ts.SyntaxKind.ExportDeclaration]: "export",
+  [ts.SyntaxKind.NamespaceExportDeclaration]: "namespace export",
+  [ts.SyntaxKind.VariableStatement]: "variable statement"
+}
+
+var firstRun = true;
+
+/**
+ * This rule prevents the exporting of extension APIs that not not meet certain release tags.
+ */
+module.exports = {
+  meta: {
+    type: "problem",
+    docs: {
+      description: "Prevent the exporting of extension APIs that do not meet certain release tags.",
+      category: "TypeScript",
+    },
+    messages: {
+      forbidden: `{{kind}} "{{name}}" without one of the release tags "{{releaseTags}}".`,
+    },
+    schema: [
+      {
+        type: "object",
+        additionalProperties: false,
+        properties: {
+          releaseTags: {
+            type: "array",
+            uniqueItems: true,
+            items: {
+              type: "string",
+              enum: ["public", "beta", "alpha", "internal", "preview"]
+            }
+          },
+          outputApiFile: {
+            type: "boolean"
+          }
+        }
+      }
+    ]
+  },
+
+  create(context) {
+    const parserServices = getParserServices(context);
+
+    const releaseTags = (context.options.length > 0 && context.options[0].releaseTags) || ["public"];
+    const extensionsTag = "extensions"; // SWB temporary extension tag name
+    const previewTag = "preview";
+
+    const outputApiFile = (context.options.length > 0 && context.options[0].outputApiFile) || false;
+    const apiFilePath = "./lib/GeneratedExtensionApi.csv";
+
+    if (firstRun) {
+      firstRun = false;
+      if (outputApiFile) {
+        // create/clear output api file on first run
+        fs.writeFileSync(apiFilePath, "");
+      }
+    }
+
+    function addToApiList(declaration, isPreview) {
+      if (!outputApiFile) {
+        return;
+      }
+
+      // Separate interfaces, items marked as real, enums, and others (treated as types)
+      let kind = "real";
```

Comment:
should we be using `syntaxKindFriendlyNames` here?

```js
kind = syntaxKindFriendlyNames[declaration.kind]
```

the values might need to get updated